controller/src/main/java/io/pravega/controller/store/stream/BucketStore.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream;
+
+import io.pravega.controller.server.bucket.BucketChangeListener;
+import io.pravega.controller.server.bucket.BucketOwnershipListener;
+import lombok.Getter;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+/**
+ * Stream Metadata.
+ */
+public interface BucketStore {
+    /**
+     * Method to get count of buckets in the store.
+     *
+     * @return number of buckets.
+     */
+    int getBucketCount();
+
+    /**
+     * Method to register listener for changes to bucket's ownership.
+     *
+     * @param serviceType service type
+     * @param listener    listener
+     */
+    void registerBucketOwnershipListener(ServiceType serviceType, BucketOwnershipListener listener);
```

Comment:
Can you register more than one listener for a service type?
If yes, then do you unregister all at once?
If no, then what is the behavior?

Same question for all register/unregister methods.

Reply:
Refactored the whole logic so now its more explicit. The register and unregister are not on the store anymore but methods on the service which makes registration and unregistration performed exactly once.